pkg/volume/csi/csi_client.go

```diff
@@ -121,10 +90,13 @@ func (c *csiDriverClient) NodePublishVolume(
 	if targetPath == "" {
 		return errors.New("missing target path")
 	}
-	if err := c.assertConnection(); err != nil {
-		glog.Errorf("%v: failed to assert a connection: %v", csiPluginName, err)
+
+	conn, err := newGrpcConn(c.driverName)
```

Comment:
nit: as a general rule of thumb never call `new` anything inside of a method implementing business logic. That encodes the creation of a dependency with the business logic making it harder to test just the business logic.

Instead, have the dependency either passed in as a parameter (with an interface type) or make the method creating the external dependency another interface method (i.e. a new `NewGrpcConnection()` method in the `csiClient interface`). That way it makes it easy to create a fake of the dependency for testing.

You don't have to do this now. Can do it in a follow up PR.

Reply:
Ok I will take a look at that.

Reply:
i looked at your suggestion. While agree, I even had an impl that injected the connection. However, the `grpc.ClientConn` type (created with newGrpcCon) is a struct value which complicates mocking it.  So, I kept it simple.  But I agree.